airframe-control/src/main/scala/wvlet/airframe/control/Parallel.scala

```diff
@@ -46,17 +65,27 @@ object Parallel {
     * @param f Function which processes each element of the source collection
     * @return Collection of the results
     */
-  def run[T, R: ClassTag](source: Seq[T], parallelism: Int = Runtime.getRuntime.availableProcessors())(
+  def run[T, R: ClassTag](source: Seq[T], parallelism: Int = Runtime.getRuntime.availableProcessors(), jmxAgent: Option[JMXAgent] = None)(
       f: T => R): Seq[R] = {
+
+    val executionId = UUID.randomUUID.toString
+    trace(s"$executionId - Begin Parallel.run (parallelism = ${parallelism})")
+
     val requestQueue = new LinkedBlockingQueue[IndexedWorker[T, R]](parallelism)
     val resultArray  = new Array[R](source.length)
 
-    Range(0, parallelism).foreach { _ =>
-      val worker = new IndexedWorker[T, R](requestQueue, resultArray, f)
+    Range(0, parallelism).foreach { i =>
+      val worker = new IndexedWorker[T, R](executionId, i.toString, requestQueue, resultArray, f)
       requestQueue.put(worker)
     }
 
     val executor = Executors.newFixedThreadPool(parallelism)
+    val counter = new AtomicLong(0)
+
+    jmxAgent.foreach { jmxAgent =>
+      val stats = new ParallelExecutionStats(executionId, parallelism, source.size.toString, requestQueue, counter)
+      jmxAgent.register(s"wvlet.airframe.control.Parallel:name=$executionId", stats)
```

Comment:
This might register too many mBeans to JVM, and will cause OOM if we frequently use Parallel.run. I think we should use a shared mBean.

Reply:
FYI. Datadog has a limitation to the number of metrics (mBeans * its parameters)we can monitor at the same time. It's around several hundreds. 

So we should report fewer metrics for this monitoring:
- How many parallel tasks are running in total?
- How many parallel workers are created in total?

Reply:
350 is Datadog's JMX metrics limit: https://docs.datadoghq.com/integrations/java/#the-350-metric-limit

Reply:
Current `Parallel` implementation is designed for running heavy tasks in parallel, not for running many parallel execution at the same. In this situation, grabbing stats of each execution accurately is better than aggregating them.

But I know that adding / removing metrics dynamically makes monitoring difficult. I would like to allow to be able to be monitored progress of each execution somehow.

Reply:
I researched JMX but I couldn't find any solution to expose metrics per execution as single MBean instance other than returning them as a structured String value like JSON. 😨 

I think aggregated metrics are not enough for current intended use cases (run heavy, but not so many elements in each execution). Do you have any idea?

Reply:
- We can use nested mBean. I think you forget to add @JMX annotation to `ParallelExecutionStats` class.

- Then I think JMX registration should be optional (instead of registering each time when `parallel` is used), and we should be able to specify MBean name if JMX monitoring is necessary. Let's talk about it later.

Reply:
I gave up to expose per execution metrics on JMX. Finally, updated to expose aggregated metrics by single static MBean instance. Also MBean registration is now optional.